### ISSUE TITLE
Include TypeScript Declaration Files

### ISSUE THREAD

**Issue by ykomatsu** [jscad/OpenJSCAD.org#403]
## Expected Behavior

OpenJSCAD.org includes updated TypeScript Declaration Files.

VS Code IntelliSence for **JavaScript** uses **TypeScript** Declation Files.

https://code.visualstudio.com/docs/languages/javascript#_intellisense

DefinetelyTyped has TypeScript Declaration Files for OpenJSCAD.org, but they look a bit old.

https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/openjscad

## Actual Behavior

OpenJSCAD.org does not include TypeScript Declaration Files.

## Steps to Reproduce the Problem

... snip ...

## Specifications

... snip ...

**Comment by z3dev** [comment]
@ykomatsu thanks for the information.

TypeScript support is not planned for OpenJSCAD, so those definitions at DefinitelyTyped were created by someone else. Maybe you can find the original author via the user group.

**Comment by danmarshall** [comment]
Hi @ykomatsu, I wrote the original typings. Since JSCAD is under active development, I have an interim set of typings on https://github.com/danmarshall/jscad-typings . It depends on which modules you are using. Here's how these are used in Maker.js: https://github.com/Microsoft/maker.js/blob/master/packages/maker.js/tsconfig.json#L7
Let me know if I can provide more info.

**Comment by z3dev** [comment]
@danmarshall Thanks again!

**Comment by ykomatsu** [comment]
@z3dev @danmarshall Thank you!

I just want to write `*.jscad` files with VS Code IntelliSense.
Probably, I need TypeScript Declaration Files of https://github.com/jscad/csg.js.

I wonder if I can convert JavaScript files to TypeScript Declaration Files, like JSDoc:

https://github.com/jscad/csg.js/blob/master/docs/api.md

**Comment by z3dev** [comment]
A few of the hard core TypeScript users are looking at this, but we don’t have the band width to work on TypeScript. Hopefully, future JavaScript standards will bring the things closer. Until then, please feel free to chat on this thread.

@danmarshall May have a few more tips for you.

**Comment by kaosat-dev** [comment]
@danmarshall yes, it is jsdoc generated, but I sadly discovered recently that is not 100% complete (blame it on the remaining mess of V1) , it will be 100% accurate once V2 is done (the actual api / what is exported will be much clearer & cleaner)

**Comment by z3dev** [comment]
@danmarshall V2 is starting to jell now. It would be nice if we could generate typescript definitions using JSDOC. I noticed that three.js is approaching typescript in the same way.

Sadly, I don’t have time to work on this. It would be good if someone with typescript knowledge and desire take this on.

**Comment by danmarshall** [comment]
@z3dev That's great news! I've been watching the Jscad repos and excited with the progress you and @kaosat-dev have been making. I'm looking forward to creating TypeScript typings for those who find them helpful, myself included! Do you think the v2 API code will be stable soon?

**Comment by z3dev** [comment]
I believe that the API for modeling is fairly stable. If there are any changes then those will be minor, i.e. name change, added option.

Should the TS definitions be added into the main repository?
Can the TS definitions be generated?

**Comment by z3dev** [comment]
@danmarshall the API for modeling is pretty much frozen now. there may be small name changes but nothing major. also, the full documntation can be generated (We are talking about putting this on-line as well).

so, maybe it's time to look at some TS typings again? can these be generated from the code?

**Comment by z3dev** [comment]
Everyone, the documentation is online now, and updated during releases. See

www.jscad.xyz/docs/

**Comment by z3dev** [comment]
Everyone!

There's a new pull request for adding TypeScript declarations. It would be really helpful if others (you) take a look at the declarations, and if possible try them.

**Comment by z3dev** [comment]
Typescript declarations are now available in V2 JSCAD.
Please try those declarations, and open a new issue for any problems found.
